### PR TITLE
Implement composite search parameters in HibernateSearch for JPA

### PR DESCRIPTION
**Pull request by michaelabuckley**

### CODE REVIEW

File: hapi-fhir-jpaserver-test-utilities/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoR4StandardQueriesLuceneTest.java [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+	@Nested
+	class CompositeSearch extends CompositeSearchParameterTestCases {
+		CompositeSearch() {
+			super(myDataBuilder, myTestDaoSearch);
+		}
+
+		/** JPA doesn't know which sub-element matches */
+		@Override
+		protected boolean isCorrelatedSupported() {
```
Comment (jmarchionatto) [inline hapi-fhir-jpaserver-test-utilities/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoR4StandardQueriesLuceneTest.java:82]: thoght: not that clear to me that not knowing which sub-element matches means supporting correlation

File: hapi-fhir-jpaserver-test-utilities/src/test/java/ca/uhn/fhir/jpa/dao/search/ExtendedHSearchIndexExtractorTest.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+import org.junit.jupiter.api.Test;
+
+import java.util.Set;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.empty;
+import static org.hamcrest.Matchers.hasSize;
+import static org.hamcrest.Matchers.not;
+
+class ExtendedHSearchIndexExtractorTest implements ITestDataBuilder.WithSupport {
```
Comment (jmarchionatto) [inline hapi-fhir-jpaserver-test-utilities/src/test/java/ca/uhn/fhir/jpa/dao/search/ExtendedHSearchIndexExtractorTest.java:27]: Looks like it needs few corner case tests

File: hapi-fhir-test-utilities/src/main/java/ca/uhn/fhir/test/utilities/ITestDataBuilder.java [hunk 3]
```diff
@@ -192,6 +202,10 @@ default IIdType createOrganization(Consumer<IBaseResource>... theModifiers) {
 	default IIdType createResource(String theResourceType, Consumer<IBaseResource>... theModifiers) {
 		IBaseResource resource = buildResource(theResourceType, theModifiers);
 
+		if (ourLog.isDebugEnabled()) {
```
Comment (jmarchionatto) [inline hapi-fhir-test-utilities/src/main/java/ca/uhn/fhir/test/utilities/ITestDataBuilder.java:205]: thought: if we updated our logging implementation, we could start using the new lambda parameter construction, which (by being lazy) doesn't require the outer condition to avoid extra work when not applicable

File: hapi-fhir-test-utilities/src/main/java/ca/uhn/fhir/test/utilities/ITestDataBuilder.java [hunk 4]
```diff
@@ -328,4 +351,58 @@ static void __setPrimitiveChild(FhirContext theFhirContext, IBaseResource theTar
...
+
+		@Override
+		default IIdType doUpdateResource(IBaseResource theResource) {
+			return getSupport().updateResource(theResource);
+		}
+	}
+
+	/**
+	 * Dummy support to use ITestDataBuilder as just a builder, not a DAO
+	 * todo mb Maybe we should split out the builder into a super-interface and drop this?
```
Comment (jmarchionatto) [inline hapi-fhir-test-utilities/src/main/java/ca/uhn/fhir/test/utilities/ITestDataBuilder.java:382]: That sounds like a good idea to me

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/search/ExtendedHSearchClauseBuilder.java [hunk 5]
```diff
@@ -65,19 +69,7 @@
...
-import static ca.uhn.fhir.jpa.model.search.HSearchIndexWriter.NUMBER_VALUE;
-import static ca.uhn.fhir.jpa.model.search.HSearchIndexWriter.QTY_CODE;
-import static ca.uhn.fhir.jpa.model.search.HSearchIndexWriter.QTY_CODE_NORM;
-import static ca.uhn.fhir.jpa.model.search.HSearchIndexWriter.QTY_PARAM_NAME;
-import static ca.uhn.fhir.jpa.model.search.HSearchIndexWriter.QTY_SYSTEM;
-import static ca.uhn.fhir.jpa.model.search.HSearchIndexWriter.QTY_VALUE;
-import static ca.uhn.fhir.jpa.model.search.HSearchIndexWriter.QTY_VALUE_NORM;
-import static ca.uhn.fhir.jpa.model.search.HSearchIndexWriter.SEARCH_PARAM_ROOT;
-import static ca.uhn.fhir.jpa.model.search.HSearchIndexWriter.URI_VALUE;
+import static ca.uhn.fhir.jpa.model.search.HSearchIndexWriter.*;
```
[the remark this answers is not in the document]
Comment (michaelabuckley) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/search/ExtendedHSearchClauseBuilder.java:72]: Yes.  My IntelliJ settings were bad.

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/search/ExtendedHSearchClauseBuilder.java [hunk 6]
```diff
@@ -384,30 +379,34 @@ public void addReferenceUnchainedSearch(String theSearchParamName, List<List<IQu
...
 	public void addDateUnmodifiedSearch(String theSearchParamName, List<List<IQueryParameterType>> theDateAndOrTerms) {
-		for (List<? extends IQueryParameterType> nextAnd : theDateAndOrTerms) {
-			// comma separated list of dates(OR list) on a date param is not applicable so grab
-			// first from default list
-			if (nextAnd.size() > 1) {
-				throw new IllegalArgumentException(Msg.code(2032) + "OR (,) searches on DATE search parameters are not supported for ElasticSearch/Lucene");
-			}
-			DateParam dateParam = (DateParam) nextAnd.stream().findFirst()
-				.orElseThrow(() -> new InvalidRequestException("Date param is missing value"));
+		for (List<? extends IQueryParameterType> nextOrList : theDateAndOrTerms) {
```
[the remark this answers is not in the document]
Comment (michaelabuckley) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/search/ExtendedHSearchClauseBuilder.java:386]: I think some earlier confusion about our AndOr lists, and all the weird data comparison operators.

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/search/ExtendedHSearchClauseBuilder.java [hunk 7]
```diff
@@ -661,4 +658,72 @@ public void addNumberUnmodifiedSearch(String theParamName, List<List<IQueryParam
...
+	}
+
+	/**
+	 * Compute the match clause for all the ocm
+	 * @param theSearchParam
+	 * @param theSubSearchParams
+	 * @param theTerm
+	 * @return
+	 */
+	private PredicateFinalStep computeCompositeTermClause(RuntimeSearchParam theSearchParam, List<RuntimeSearchParam> theSubSearchParams, CompositeParam theTerm) {
```
[the remark this answers is not in the document]
Comment (michaelabuckley) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/search/ExtendedHSearchClauseBuilder.java:686]: I'd love that.  Unfortunately, PredicateFinalStep and all the HSearch apis are completely opaque.  Can't inspect them at all.

File: hapi-fhir-jpaserver-searchparam/src/main/java/ca/uhn/fhir/jpa/searchparam/extractor/BaseSearchParamExtractor.java [hunk 8]
```diff
@@ -211,7 +211,7 @@ public List<String> extractParamValuesAsStrings(RuntimeSearchParam theSearchPara
 				return extractReferenceParamsAsQueryTokens(theSearchParam, theResource, extractor);
 			case QUANTITY:
 				if (myModelConfig.getNormalizedQuantitySearchLevel().equals(NormalizedQuantitySearchLevel.NORMALIZED_QUANTITY_SEARCH_SUPPORTED)) {
```
[the remark this answers is not in the document]
Comment (michaelabuckley) [reply hapi-fhir-jpaserver-searchparam/src/main/java/ca/uhn/fhir/jpa/searchparam/extractor/BaseSearchParamExtractor.java:227]: Nice.  Done.

File: hapi-fhir-jpaserver-model/src/main/java/ca/uhn/fhir/jpa/model/search/ExtendedHSearchIndexData.java [hunk 9]
```diff
@@ -94,11 +99,11 @@ public void writeIndexElements(DocumentElement theDocument) {
 		mySearchParamTokens.forEach(ifNotContained(indexWriter::writeTokenIndex));
 		mySearchParamLinks.forEach(ifNotContained(indexWriter::writeReferenceIndex));
 		// we want to receive the whole entry collection for each invocation
-		Multimaps.asMap(mySearchParamQuantities).forEach(ifNotContained(indexWriter::writeQuantityIndex));
+		mySearchParamQuantities.forEach(ifNotContained(indexWriter::writeQuantityIndex));
```
[the remark this answers is not in the document]
Comment (michaelabuckley) [reply hapi-fhir-jpaserver-model/src/main/java/ca/uhn/fhir/jpa/model/search/ExtendedHSearchIndexData.java:101]: It is the reverse, in fact.  Each quantity should be it's own separate nested document, or the unit+value correlation will be muddled.

[the remark this answers is not in the document]
Comment (jmarchionatto) [reply hapi-fhir-jpaserver-model/src/main/java/ca/uhn/fhir/jpa/model/search/ExtendedHSearchIndexData.java:101]: But how was then working, if we were always storing quantity in a nested structure. Isn't elastic automatically separating them different documents because of being nested?

[the remark this answers is not in the document]
Comment (michaelabuckley) [reply hapi-fhir-jpaserver-model/src/main/java/ca/uhn/fhir/jpa/model/search/ExtendedHSearchIndexData.java:101]: You are correct.  I've gone back to passing in the context factory.

File: hapi-fhir-jpaserver-test-utilities/src/main/java/ca/uhn/fhir/jpa/search/QuantitySearchParameterTestCases.java [hunk 10]
```diff
@@ -0,0 +1,634 @@
...
+	}
+
+	@Override
+	public Support getTestDataBuilderSupport() {
+		return myTestDataBuilder;
+	}
+
+
+	@Nested
+		public class QuantitySearch {
```
Comment (TipzCM) [inline hapi-fhir-jpaserver-test-utilities/src/main/java/ca/uhn/fhir/jpa/search/QuantitySearchParameterTestCases.java:52]: the spacing 😣

File: hapi-fhir-jpaserver-test-utilities/src/main/java/ca/uhn/fhir/jpa/search/QuantitySearchParameterTestCases.java [hunk 11]
```diff
@@ -0,0 +1,634 @@
...
+				public void ne() {
+					withObservationWithValueQuantity(0.6);
+
+					assertFind("when gt", "/Observation?value-quantity=ne0.5");
+					assertNotFind("when eq", "/Observation?value-quantity=ne0.6");
+					assertFind("when lt", "/Observation?value-quantity=ne0.7");
+				}
+
+				@Test
+				public void ap() {
```
Comment (TipzCM) [inline hapi-fhir-jpaserver-test-utilities/src/main/java/ca/uhn/fhir/jpa/search/QuantitySearchParameterTestCases.java:99]: what does ap mean?

**Review by TipzCM — APPROVED** [review]

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/search/ExtendedHSearchClauseBuilder.java [hunk 12]
```diff
@@ -138,17 +130,21 @@ private Set<String> extractOrStringParams(List<? extends IQueryParameterType> ne
...
+	 *
+	 * Wrap the predicates under a bool as should clauses with minimumShouldMatch=1 for OR semantics.
+	 * As an optimization, when there is only one clause, we avoid the redundant boolean wrapper
+	 * and return the first item as is.
 	 *
 	 * @param theOrList a list containing at least 1 predicate
-	 * @return a predicate providing or-sematics over the list.
+	 * @return a predicate providing or-semantics over the list.
 	 */
 	private PredicateFinalStep orPredicateOrSingle(List<? extends PredicateFinalStep> theOrList) {
```
[the remark this answers is not in the document]
Comment (michaelabuckley) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/search/ExtendedHSearchClauseBuilder.java]: Thanks, Juan.  I had to put all the nested ones back.  I now understand that the root predicate builder can't be used to build predicates inside nested clauses.

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/search/ExtendedHSearchClauseBuilder.java [hunk 13]
```diff
@@ -509,55 +508,54 @@ private PredicateFinalStep generateDateInstantSearchTerms(String theSearchParamN
...
+				String valueFieldPath = quantityElement + "." + QTY_VALUE_NORM;
+				setPrefixedNumericPredicate(quantityClause, activePrefix, canonicalQty.getValue(), valueFieldPath, true);
+				quantityClause.must(myPredicateFactory.match()
+					.field(quantityElement + "." + QTY_CODE_NORM)
 					.matching(canonicalQty.getUnits()));
-				return;
+				finished = true;
 			}
 		}
+		if (!finished) {// not NORMALIZED_QUANTITY_SEARCH_SUPPORTED or non-canonicalizable parameter
```
[the remark this answers is not in the document]
Comment (michaelabuckley) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/search/ExtendedHSearchClauseBuilder.java]: An artifact of refactoring tools.  Gone.

File: hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/docs/server_jpa/elastic.md [hunk 14]
```diff
@@ -7,81 +7,77 @@ This is required to support the `_content`, or `_text` search parameters.
...
 [Lucene](https://lucene.apache.org/core/8_10_1/queryparser/org/apache/lucene/queryparser/simple/SimpleQueryParser.html) and
 [Elasticsearch](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-simple-query-string-query.html#simple-query-string-syntax).
 Terms are delimited by whitespace, or query punctuation `"()|+'`.  Literal uses of these characters must be escaped by '\'.
 When the query only contains one or more bare terms, they are each converted to a prefix search to match the behaviour of a normal string search.
-When multiple terms are present, they must all match (i.e. `AND`).  For `OR` behaviour use the `|` operator between terms. 
-But if any special SQS syntax is active, the query is used as is.
-To ensure that the query is used as-is, quote bare terms with the `"` or  character.  E.g. `without any special syntax characters 
+When multiple terms are present, they must all match (i.e. `AND`).  For `OR` behaviour use the `|` operator between terms.
+If any special SQS syntax is active, the query is used as is.
+To ensure that the query is used as-is, quote bare terms with the `"` or  character.  I.e. without any special syntax characters.
```
[the remark this answers is not in the document]
Comment (michaelabuckley) [reply hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/docs/server_jpa/elastic.md:49]: Thanks.  Fixed and re-wrote some of this.

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/search/ExtendedHSearchClauseBuilder.java [hunk 15]
```diff
@@ -384,30 +394,34 @@ public void addReferenceUnchainedSearch(String theSearchParamName, List<List<IQu
...
-			myRootClause.must(searchPredicate);
+			myRootClause.must(orPredicateOrSingle(clauses, myRootPredicateFactory));
 		}
 	}
 
-	private PredicateFinalStep generateDateOrdinalSearchTerms(String theSearchParamName, DateParam theDateParam) {
-		String lowerOrdinalField = SEARCH_PARAM_ROOT + "." + theSearchParamName + ".dt.lower-ord";
-		String upperOrdinalField = SEARCH_PARAM_ROOT + "." + theSearchParamName + ".dt.upper-ord";
+	private PredicateFinalStep buildDateTermClause(String thePath, IQueryParameterType theParam) {
+		DateParam dateParam = (DateParam) theParam;
```
[the remark this answers is not in the document]
Comment (michaelabuckley) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/search/ExtendedHSearchClauseBuilder.java:414]: Consistency of api.  Some of the other clauses can consume multiple types, so the root class is passed.

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/search/ExtendedHSearchClauseBuilder.java [hunk 16]
```diff
@@ -509,61 +523,76 @@ private PredicateFinalStep generateDateInstantSearchTerms(String theSearchParamN
...
+
+					return orPredicateOrSingle(orClauses, nestedRootPredicateFactory);
+					});
 
-			myRootClause.must(quantityTerms);
+			myRootClause.must(orListPredicate);
 		}
 	}
 
+	private BooleanPredicateClausesStep<?> buildQuantityTermClause(String spPath, IQueryParameterType paramType, SearchPredicateFactory theContextPredicateFactory) {
```
[the remark this answers is not in the document]
Comment (michaelabuckley) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/search/ExtendedHSearchClauseBuilder.java:543]: Yup.

File: hapi-fhir-jpaserver-searchparam/src/main/java/ca/uhn/fhir/jpa/searchparam/extractor/BaseSearchParamExtractor.java [hunk 17]
```diff
@@ -258,6 +265,126 @@ private <T extends BaseResourceIndexedSearchParam> List<String> toStringList(Sea
...
+		 * Extract the subcomponent index data for each component of a composite SP from an IBase element.
+		 *
+		 * @param theParams               will add 1 or 0 ResourceIndexedSearchParamComposite instances for theValue
+		 * @param theCompositeSearchParam the composite SP
+		 * @param theValue                the focus element for the subcomponent extraction
+		 * @param thePath                 unused api param
+		 * @param theWantLocalReferences  passed down to reference extraction
+		 */
+		@Override
+		public void extract(SearchParamSet<ResourceIndexedSearchParamComposite> theParams, RuntimeSearchParam theCompositeSearchParam, IBase theValue, String thePath, boolean theWantLocalReferences) {
```
[the remark this answers is not in the document]
Comment (michaelabuckley) [reply hapi-fhir-jpaserver-searchparam/src/main/java/ca/uhn/fhir/jpa/searchparam/extractor/BaseSearchParamExtractor.java:303]: I like the idea.  I'll do that in the future.

File: hapi-fhir-jpaserver-test-utilities/src/main/java/ca/uhn/fhir/jpa/search/QuantitySearchParameterTestCases.java (hunk 11 above)
Comment (michaelabuckley) [reply hapi-fhir-jpaserver-test-utilities/src/main/java/ca/uhn/fhir/jpa/search/QuantitySearchParameterTestCases.java:99]: Approximate.  This is a standard fhir prefix.  e.g. `Observation?value-quantity=ap8.2` means approximately (+- 10%) of 8.2

File: hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoR4StandardQueriesNoFTTest.java [hunk 18]
```diff
@@ -410,4 +413,25 @@ public void sortByNumeric() {
 
 	}
 
+	// wipmb re-enable this.  Some of these fail!
```
[the remark this answers is not in the document]
Comment (michaelabuckley) [reply hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoR4StandardQueriesNoFTTest.java:416]: I'll log it.

File: hapi-fhir-jpaserver-searchparam/src/main/java/ca/uhn/fhir/jpa/searchparam/extractor/BaseSearchParamExtractor.java [hunk 19]
```diff
@@ -258,6 +265,126 @@ private <T extends BaseResourceIndexedSearchParam> List<String> toStringList(Sea
...
+	private IExtractor<ResourceIndexedSearchParamComposite> createCompositeExtractor(IBaseResource theResource) {
+		return new CompositeExtractor(theResource);
+
+	}
+
+	/**
+	 * Extractor for composite SPs.
+	 * Extracts elements, and then recurses and applies extractors for each component SP using the element as the new root.
+	 */
+	public class CompositeExtractor implements IExtractor<ResourceIndexedSearchParamComposite> {
```
[the remark this answers is not in the document]
Comment (michaelabuckley) [reply hapi-fhir-jpaserver-searchparam/src/main/java/ca/uhn/fhir/jpa/searchparam/extractor/BaseSearchParamExtractor.java:284]: Next time.  Agree that they are all tangled.  But I'd want to do all of them, and this is already a pretty big commit.

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/search/ExtendedHSearchClauseBuilder.java [hunk 20]
```diff
@@ -661,4 +694,96 @@ public void addNumberUnmodifiedSearch(String theParamName, List<List<IQueryParam
...
+	 * @param theContextPredicateFactory
+	 * @return
+	 */
+	private PredicateFinalStep computeCompositeTermClause(RuntimeSearchParam theSearchParam, List<RuntimeSearchParam> theSubSearchParams, CompositeParam theTerm, SearchPredicateFactory theContextPredicateFactory) {
+		Validate.notNull(theSearchParam);
+		Validate.notNull(theSubSearchParams);
+		Validate.notNull(theTerm);
+		Validate.isTrue(theSubSearchParams.size() == 2, "Hapi only supports composite search parameters with 2 components. %s %d", theSearchParam.getName(), theSubSearchParams.size());
+		List<IQueryParameterType> values = theTerm.getValues();
+		Validate.isTrue(theSubSearchParams.size() == values.size(), "Different number of query components than defined. %s %d %d", theSearchParam.getName(), theSubSearchParams.size(), values.size());
```
[the remark this answers is not in the document]
Comment (michaelabuckley) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/search/ExtendedHSearchClauseBuilder.java:742]: Yes.  In fact, R4 defines some standard params with 3 components.

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/search/HSearchCompositeSearchIndexDataImpl.java [hunk 21]
```diff
@@ -0,0 +1,158 @@
...
+						.flatMap(o->ObjectUtil.castIfInstanceof(o, ResourceIndexedSearchParamString.class).stream())
+						.forEach(risps-> theHSearchIndexWriter.writeBasicStringFields(stringElement, risps.getValueExact()));
+					break;
+
+				case TOKEN:
+					DocumentElement tokenElement =  subParamElement.addObject("token");
+					subParam.getParamIndexValues().stream()
+						.flatMap(o->ObjectUtil.castIfInstanceof(o, ResourceIndexedSearchParamToken.class).stream())
+						.forEach(rispt-> theHSearchIndexWriter.writeTokenFields(tokenElement, new Tag(rispt.getSystem(), rispt.getValue())));
+				break;
```
[the remark this answers is not in the document]
Comment (michaelabuckley) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/search/HSearchCompositeSearchIndexDataImpl.java:129]: It's just wrong.  Fixed.
